### PR TITLE
[fuzz] Add a meta-differential fuzz target

### PR DESCRIPTION
**Pull request by abrown**
This fuzz target combines past efforts to fuzz Wasmtime against various engines and against various kinds of modules. Once complete, it is intended to replace the other differential targets. Some notable changes in this PR include the addition of interfaces for each Wasm engine we want to fuzz against and the ability to evaluate a function multiple times with different randomly-generated inputs. See the commit messages for more details.

### CODE REVIEW

**Comment by jameysharp** [comment]
Ooh, I am eager to look at this on Monday!

File: crates/fuzzing/src/generators/value.rs
Comment (alexcrichton) [inline crates/fuzzing/src/generators/value.rs:81]: One thing that might be good to add here is "known values" for each lane width like (1u64, 1u64), (-1i64, -1i64), (1u32, 2u32, 3u32, 4u32), etc.

File: crates/fuzzing/src/generators/value.rs
Comment (alexcrichton) [inline crates/fuzzing/src/generators/value.rs:144]: This might be better modeled as `TryFrom` to have callers acknowledge that the `From` is fallible.

File: crates/fuzzing/src/generators/module.rs
Comment (alexcrichton) [inline crates/fuzzing/src/generators/module.rs:100]: Instead of having a separate structure for this couldd all of `ModuleConfig` be passed around? Otherwise this seems a bit odd since it's missing a number of other proposals which could perhaps be relevant.

File: fuzz/fuzz_targets/differential-new.rs
Comment (alexcrichton) [inline fuzz/fuzz_targets/differential-new.rs:58]: This loop may not play nicely with fuel right now since if the module has wasm-smith-injected fuel the fuel is never reset and otherwise with Wasmtime-injected fuel I'm not sure that's ever reset either.

File: fuzz/fuzz_targets/differential-new.rs
Comment (alexcrichton) [inline fuzz/fuzz_targets/differential-new.rs:11]: I think if this is `&[u8]` instead of `[u8; 1024]` it's whatever the fuzzer gives, and otherwise it's required to be 1kb?

File: crates/fuzzing/src/oracles/engine.rs
Comment (alexcrichton) [inline crates/fuzzing/src/oracles/engine.rs:66]: Reading over the implementations of this could this trait perhaps get changed to have something different like "get the hashable export named `foo`" to avoid requiring hardcoding the hash algorithm between all engines? That would also allow more direct comparisons and better errors along the lines of "the state of this global/memory is different" rather than "some state is different"

File: crates/fuzzing/src/oracles/diff_wasmtime.rs
Comment (alexcrichton) [inline crates/fuzzing/src/oracles/diff_wasmtime.rs:130]: Could this be `&mut self` to avoid the `RefCell` needed for Wasmtime?

File: crates/fuzzing/src/oracles/diff_wasmtime.rs
Comment (alexcrichton) [inline crates/fuzzing/src/oracles/diff_wasmtime.rs:32]: Personally I get pretty easily lost in the maze of `Arbitrary` implementations and methods of configuration. Would it be possible to avoid having this and `arbitrary_with_compatible_config` below? Could some higher-up generate the `Config`/`Engine` and be responsible for passing it down?

Or otherwise it seems like with Wasmtime we'd want a constant `ModuleConfig` with possibly differing `WasmtimeConfig` or something like that.

If this is too difficult though or otherwise requires complications then I think that removing differential-wasmtime-with-itself is pretty reasonable. 90% of the value comes from differential execution with _something_ so adding more engines IMO doesn't increase our coverage really all that much.

Comment (alexcrichton) [reply crates/fuzzing/src/oracles/diff_wasmtime.rs:32]: Although as I say that I remember that different engines implement different features which is often the hard part and nothing implements Wasmtime's feature set other than Wasmtime exactly.

File: crates/fuzzing/src/oracles/diff_wasmtime.rs
Comment (alexcrichton) [inline crates/fuzzing/src/oracles/diff_wasmtime.rs:181]: I think that this is worth implementing but all that we can really say about these is "null" or "non-null" so I think we could have `DiffValue::ExternRef { is_null: bool }` perhaps

File: fuzz/fuzz_targets/differential-new.rs
Comment (alexcrichton) [inline fuzz/fuzz_targets/differential-new.rs:54]: One thing that may be helpful here is to use the concrete Wasmtime rhs to use Wasmtime's embedding API to determine this rather than using the `wasmparser` crate. 

I also think it's reasonable for the `differential` function itself to just force the rhs to the Wasmtime concrete type because I don't think there's much value in having two 100% generic engines since we always want one to be Wasmtime anyway.

**Review by alexcrichton — COMMENTED** [review]
This looks great to me! I think there's cleanup to do internally and such but otherwise the overall architecture/design looks pretty reasonable. I'd like to ideally simplify the Wasmtime bits a bit since there seems to be good deal of juggling various bits of configuration.

I'm happy to help implement the v8 bits as well.

File: crates/fuzzing/src/oracles/diff_spec.rs
Comment (fitzgen) [inline crates/fuzzing/src/oracles/diff_spec.rs:94]: This stuff should probably be in `generators/value.rs`.

File: crates/fuzzing/src/oracles/diff_wasmi.rs
Comment (fitzgen) [inline crates/fuzzing/src/oracles/diff_wasmi.rs:150]: Ditto

File: fuzz/fuzz_targets/differential-new.rs
Comment (fitzgen) [inline fuzz/fuzz_targets/differential-new.rs:63]: Running the exported functions multiple times will -- in general, between single-instruction modules and fuel limits -- be a lot cheaper than compiling and instantiating the module itself. Therefore, we should make sure we take as much advantage of the work we already did to get here by at least using each of our known values for each argument, and then whatever the fuzzer gives us that is left over can also be used to generate new inputs, but at least this way we know we are exercising every module with a bunch of different interesting argument values.

File: fuzz/fuzz_targets/differential-new.rs
Comment (fitzgen) [reply fuzz/fuzz_targets/differential-new.rs:58]: Yeah, we should reset fuel for each call.

**Review by fitzgen — COMMENTED** [review]
Super excited for this! Thanks for taking this on, Andrew!

File: crates/fuzzing/src/oracles/diff_wasmtime.rs
Comment (abrown) [reply crates/fuzzing/src/oracles/diff_wasmtime.rs:130]: It could... but actually it feels like the Wasmtime API shouldn't require a mutable store when we're only reading values.

Comment (alexcrichton) [reply crates/fuzzing/src/oracles/diff_wasmtime.rs:130]: Wasmtime has its own internal reasons for this, namely that extracting items from an instance is done lazily and may insert items into the store. There's other possibilities like reading an externref global could trigger a gc or similar situations like that. 

Overall I think it's best to stick with `&mut` here.

**Comment by jameysharp** [comment]
It's been suggested (by a couple different people, I think, but I've lost track of who) that when we have a wasm program we want to try running, we should run it with every engine we have, not just a pair of engines.

I don't think that needs to happen in this PR. I just wanted to note that if we do want that, I think the changes I see here already cover most of the work that would be needed. So that's great!

One of my teammates also suggested that we should check that different engines agree on whether a random sequence of bytes is a valid wasm binary. That seems like another relatively small extension of this work, that somebody could build after this lands.

**Comment by alexcrichton** [comment]
One worry I have about running against all engines is that we run the risk of really hurting executions per second during fuzzing. When something fails I think it might be good to run in other engines to figure out which answer is probably correct, or otherwise being able to run in multiple engines easily locally would be good. While fuzzing though I think it's best to stick to just 2 engines with one guaranteed as Wasmtime

File: Cargo.lock
Comment (jameysharp) [inline Cargo.lock:92]: Is this newer version of `anyhow` necessary? If so, I think it should probably be reflected in the minimum version requirement in the appropriate `Cargo.toml`. Otherwise can you revert this change to `Cargo.lock`?

If you do need the upgrade, I don't think we should be adding cargo-vet exceptions. I hope a delta certification isn't a big deal for that crate.

If you're just adding the exception as a temporary thing to get CI past that check, could you clearly mark that commit so we don't forget to remove it again before merging?

Comment (abrown) [reply Cargo.lock:92]: No, sorry. I was just trying to make the CI green but was not reflecting too much on why this was even necessary. I think at one point in the lifetime of this PR I had brought anyhow to the fuzz targets but that is no longer necessary (besides, someone else has already brought that dependency in since this PR was opened). I'll fix up the lock file instead.

Comment (jameysharp) [reply Cargo.lock:92]: That's totally understandable!

On the other PR where you (briefly) added this exemption, I was worried we'd merged a lockfile change to `main` that somehow slipped past the cargo-vet CI checks. That's what I meant by wondering if we'd screwed up: it wouldn't be hugely surprising if either cargo-vet or the CI integration had a bug, and besides, as a community our processes around this are largely untested still. But it's nice to know that this particular case has a simpler explanation.

Comment (abrown) [reply Cargo.lock]: Well, thanks for catching this (and the other PR!). This commit should be gone now as well as the unintential `Cargo.lock` upgrade.

File: crates/fuzzing/src/generators/module.rs
Comment (abrown) [reply crates/fuzzing/src/generators/module.rs:100]: I initially designed it like that but then moved to this when I realized that `SwarmConfig` (the `wasm-smith` struct what `ModuleConfig` wraps) is intended to "generate a module this way" and what I mean is "does a module use this feature?" So a lot of `SwarmConfig`, e.g., the minimum/maximums of various kinds, don't really apply. They don't matter to the engines for the most part (an engine can handle 10 instructions just as well as 100 instructions) and they don't make sense for the single-instruction modules (they always only have 1 instruction). That's when I switched from adapting a single-instance module to a `ModuleConfig/SwarmConfig` and instead started adapting a `ModuleConfig/SwarmConfig` to this new `ModuleFeatures`. Would you be more in favor of this struct if I added the remaining Wasm features to it?

File: fuzz/fuzz_targets/differential-new.rs
Comment (abrown) [reply fuzz/fuzz_targets/differential-new.rs]: I don't see how this would change things much: we still want a separate function to gather up the signatures of the functions to execute. `wasmparser` is already a dependency and it doesn't seem to me that switching to Wasmtime would reduce much code (?).

File: crates/fuzzing/src/oracles/diff_spec.rs
Comment (abrown) [reply crates/fuzzing/src/oracles/diff_spec.rs:94]: I was dividing the code up by engine, so I chose to put all these engine-specific conversions together with the engine that needs them.

**Comment by abrown** [comment]
All right, I'm marking this ready for review with some caveats that I think should be addressed in subsequent PRs. At this point, the PR provides a new interface for evaluating various engines against Wasmtime but for now is only is using single-instruction modules to do so. After resolving a bunch of issues that this identified (all related to fuzz infrastructure issues, IIRC), I was able to run the target for ~3 million test cases without crashing:

```
#2963550        REDUCE cov: 30985 ft: 87395 corp: 1762/219Kb lim: 4096 exec/s: 394 rss: 676Mb L: 140/521 MS: 1 EraseBytes-
ignoring error: this error should be ignored by fuzzing: unable to compile module in wasmtime
ignoring error: this error should be ignored by fuzzing: unable to compile module in wasmtime
ignoring error: this error should be ignored by fuzzing: unable to compile module in wasmtime
ignoring error: this error should be ignored by fuzzing: Unknown opcode 195
ignoring error: this error should be ignored by fuzzing: Unknown opcode 195
ignoring error: this error should be ignored by fuzzing: Unknown opcode 195
ignoring error: this error should be ignored by fuzzing: Unknown opcode 195
ignoring error: this error should be ignored by fuzzing: unable to compile module in wasmtime
ignoring error: this error should be ignored by fuzzing: unable to compile module in wasmtime
ignoring error: this error should be ignored by fuzzing: unable to compile module in wasmtime
ignoring error: this error should be ignored by fuzzing: unable to compile module in wasmtime
ignoring error: this error should be ignored by fuzzing: unable to compile module in wasmtime
ignoring error: this error should be ignored by fuzzing: unable to compile module in wasmtime
ignoring error: this error should be ignored by fuzzing: unable to compile module in wasmtime
#2964099        REDUCE cov: 30985 ft: 87395 corp: 1762/219Kb lim: 4096 exec/s: 394 rss: 676Mb L: 131/521 MS: 4
PersAutoDict-ShuffleBytes-ShuffleBytes-EraseBytes- DE: "\x17\x00\x00\x00\x00\x00\x00\x00"-
ignoring error: this error should be ignored by fuzzing: unable to compile module in wasmtime
```
Note how ignored errors are printed but don't cause a crash: `wasmi` doesn't understand all of the opcodes (which probably means we need to make the single-inst modules smarter to report when they use float-to-int conversions, e.g.) and Wasmtime frequently crashes when the pooling allocator picks sizes that are too small.

When run with `RUST_LOG=wasmtime_fuzzing=debug cargo +nightly fuzz run differential_meta`, the log output shows something like:

```
[2022-08-11T20:36:44Z DEBUG wasmtime_fuzzing::oracles] wrote wasm file to `testcase14.wasm`
[2022-08-11T20:36:44Z DEBUG wasmtime_fuzzing::oracles] Evaluating: test([F32(198)])
[2022-08-11T20:36:44Z DEBUG wasmtime_fuzzing::oracles]  -> results on wasmi: Ok([F32(0)])
[2022-08-11T20:36:44Z DEBUG wasmtime_fuzzing::oracles]  -> results on wasmtime: Ok([F32(0)])
[2022-08-11T20:36:44Z DEBUG wasmtime_fuzzing::oracles] Hashing instances:
[2022-08-11T20:36:44Z DEBUG wasmtime_fuzzing::oracles]  -> hash of wasmi: 15130871412783076140
[2022-08-11T20:36:44Z DEBUG wasmtime_fuzzing::oracles]  -> hash of wasmtime: 15130871412783076140
```

I expect that once the following big items are finished we could remove all other differential targets and just use this one. For now, though, I would like to get this work merged so I can stop rebasing. This is work that I think could go in different PRs (but I am certainly open for discussion on that):
- [ ] adapt V8 to the `DiffEngine`/`DiffInstance` interface (@alexcrichton volunteered to look into this; the lifetime issues made this a bridge too far for this PR)
- [ ] start fuzzing with the `wasm-smith`-generated modules (`BigModule`) instead of just single-instruction modules; this is complicated by the fact that `wasm-smith` will eat up all remaining unstructured data, leaving none for choosing engines and function arguments
- [ ] figure out whether to fully commit to `ModuleFeatures` and report what Wasm feature a single-instruction uses (see comments above about this)
- [ ] refactor how we ignore errors: I currently do this by wrapping the error message in `DiffIgnoreError` and downcasting the `anyhow` error to see if it is one of these; @alexcrichton mentioned that he preferred `Result<Option<...>>` to indicate the various states but this would involve some refactoring
- [ ] add more instructions to the single-instruction module generator; e.g., my next feature to add would be SIMD instructions, which would likely find some bugs and require some additional infrastructure (e.g., V128 relaxed NaN comparison)

**Comment by abrown** [comment]
Oh, one other point: I have gone back and forth a few times about how to design the interface and today, at least, I'm leaning towards merging the `DiffEngine` and `DiffInstance` traits. Initially I separated them because logically that's how we're used to thinking about these and it is nice to be able filter out engines that won't work with a certain module. I feel like there is a way to just use the `DiffInstance::instantiate` function to do all of this. This refactoring change doesn't change the essence of the PR above, though. On a related note, the hashing interface could also change: I've been thinking through how hashing could work in the OCaml spec interpreter and, once I figure out the OCaml-Rust bindings issues, it may be that the interface ends up looking like what @alexcrichton [suggested above](https://github.com/bytecodealliance/wasmtime/pull/4515#discussion_r928968551).

File: crates/fuzzing/src/oracles/diff_wasmtime.rs
Comment (abrown) [reply crates/fuzzing/src/oracles/diff_wasmtime.rs:131]: The single-instruction modules that this PR uses don't exercise reference types but I agree that in the future, e.g., with `wasm-smith`-generated modules, this is probably what we want to do.

File: fuzz/fuzz_targets/differential-new.rs
Comment (abrown) [reply fuzz/fuzz_targets/differential-new.rs]: Right now, with only single-instruction modules, this shouldn't be too much of an issue but once we add in the wasm-smith modules this will be necessary; what is the right way to reset this?

File: crates/fuzzing/src/oracles/engine.rs
Comment (abrown) [reply crates/fuzzing/src/oracles/engine.rs]: Yeah, this probably does need some changes but I'm not yet entirely clear what they are; I think looking at what is necessary for the OCaml spec interpreter might make things more clear (as mentioned [here](https://github.com/bytecodealliance/wasmtime/pull/4515#issuecomment-1212486350)).

File: crates/fuzzing/src/generators/module.rs
Comment (alexcrichton) [reply crates/fuzzing/src/generators/module.rs:100]: I ideally want to centralize on feature management instead of having multiple avenues and tests in different places for features. Fuzzing all the various features gets nontrivial in management complexity pretty quickly and I always get lost a few lines in whenever I try to bring myself back up to speed on the fuzzers.

Is it a problem to take `&ModuleConfig` instead of `&ModuleFeatures`? I presume that something motivated you to add this but I don't know what that is. I'd ideally like to tackle this problem if we can.

File: fuzz/fuzz_targets/differential-new.rs
Comment (alexcrichton) [reply fuzz/fuzz_targets/differential-new.rs]: Currently there's no way to reset the fuel, we'd have to add an exported function in the `wasm-smith` generated module (or export the global) for it to get modified by the embedder.

File: fuzz/fuzz_targets/differential-new.rs
Comment (alexcrichton) [reply fuzz/fuzz_targets/differential-new.rs]: Using the Wasmtime embedding API is much more natural in my opinion, and the Wasmtime API is much more stable than the `wasmparser` API. I don't really want to keep up with random `wasmparser` changes way over here in the fuzzing code when one of the engines we're fuzzing against should always be Wasmtime anyway.

(I also don't think it's useful to be so abstract here that we could fuzz v8 against wasmi here, I would rather have "here's the wasmtime config" and "here's the other engine config" and the fuzzing is entirely driven by the Wasmtime embedding API plus requests to the engine config through a trait)

**Review by alexcrichton — COMMENTED** [review]
I'm ok setting some things aside for future PRs, but at the same time I want to make sure that what's landed here is a solid base to build these future features on to. For example I don't think that `ModuleFeatures` is going to play well over time as keeping that in sync with `ModuleConfig` and the myriad of other possible configuration options. Instead I think it's best to simplify the configuration of the module generation as much as possible.

I think a better idea would be to structure the fuzzer as:

* Generate a module configuration
* Call `set_differential_config` to ensure that the module exhibits runtime behavior independent of the engine being run within (e.g. maximum memory limits, nan canonicalization, etc). I think that feature-related options in `set_differential_config` should all get removed, however.
* Select a second engine to fuzz against with this module config. The config would act as a form of filter where if reference types are enabled then `wasmi` would be disabled since it doesn't implement reference types. Using wasmtime as a second engine is always a possibility though.
* Afterwards the module is generated and instantiated in both engines. 
* Using Wasmtime as fixed-to-one-side the embedding API is used to explore the API surface area of the module. For N iterations exported functiosn are selected at random and then given arbitrary values, with the results being tested aftewards.

I think this structure would fix some of my concerns related to relying on `wasmparser` and/or otherwise trying to avoid the Wasmtime embedding API, avoiding multiple layers of configurations, filters, etc to get code that can run within each engine, and also additionally retaining maximum flexibility to run a differential config between different engines with as many shapes of modules as possible.

**Comment by abrown** [comment]
Ok, I've rebased this PR and included some work @alexcrichton and I did today to adapt the fuzz target to his comments above (thanks @alexcrichton!). There are still some TODOs, noted in the commit message, but this now runs without error for a decent amount of time:

```
$ cargo +nightly fuzz run differential_meta -s none
...
=== Execution rate (464537 successes / 880000 attempted modules): 52.788295454545455% (total invocations: 3525516) ===
#880057 REDUCE cov: 28361 ft: 150473 corp: 9884/4009Kb lim: 627 exec/s: 531 rss: 174Mb L: 546/626 MS: 1 EraseBytes-
#880293 REDUCE cov: 28361 ft: 150473 corp: 9884/4009Kb lim: 627 exec/s: 530 rss: 174Mb L: 332/626 MS: 1 EraseBytes-
#880511 NEW    cov: 28361 ft: 150475 corp: 9885/4010Kb lim: 627 exec/s: 531 rss: 174Mb L: 622/626 MS: 3 CMP-CrossOver-ChangeByte- DE: "\x08\xa2"-
```

**Comment by alexcrichton** [comment]
Thanks again for being patient with me here @abrown! I'm happy with where this has ended up and while I think there's more to do I think it's safe to do that all as a follow-up. I'm gonna merge this and see if we can get this into the next oss-fuzz build to get our first (potentially empty?) crop of fuzz bugs.

**Review by alexcrichton — APPROVED** [review]

**Comment by abrown** [comment]
@alexcrichton, thanks for all the help reviewing this and reworking some of the bits!

**Pull request merged**
